File: elasticmeasures/Dep/dtw_scb.py
def dtw_scb(x,y,w=5):

    cur = np.full(len(y),np.inf);
    prev = np.full(len(y),np.inf);
    for i in range(len(x)):
        minw = max(0,i-w);
        maxw = min(len(y),w);
        temp = prev;
        prev = cur;
        cur = temp;
        for j in range(int(minw),int(maxw)):

            if i + j == 0:
                cur[j] = abs(x[0] - y[0]) ** 2;
            elif i == 0:
                cur[j] = abs(x[0] - y[j]) ** 2 + cur[j-1];
            elif j == 0:
                cur[j] = abs(x[i] - y[0]) ** 2 + prev[j];
            else:
                cur[j] = abs(x[i] - y[j]) ** 2 + min(prev[j-1],prev[j],cur[j-1]);
    final_dtw = cur[len(y)-1];

    return final_dtw ** (1/2);

File: elasticmeasures/Dep/dtw_scb.cpp
#include <stdlib.h>
#include <stdio.h>
#include <math.h>


#define min(x, y) ((x)<(y)?(x):(y))
#define max(x, y) ((x)>(y)?(x):(y))
#define dist(x, y)((x-y)*(x-y))



double dtw_scb(double* x, int xlen, double* y, int ylen, int w)
{

	double* prev = (double*) malloc(sizeof(double*) * ylen);
	double* cur = (double*) malloc(sizeof(double*) * ylen);
	double* temp;
	int i, j;
	for (i = 0; i < xlen; i++)
	{
		temp = prev;
		prev = cur;
		cur = temp;
		for (j = max(0, i - w); j < min(ylen, i + w); j++)
		{
			if (i + j == 0)
			{
				cur[j] = dist(x[i], y[j]);
			}
			else if (i == 0)
			{
				cur[j] = dist(x[0], y[j]) + cur[j - 1];
			}
			else if (j == 0)
			{
				cur[j] = dist(x[i], y[0]) + prev[j];
			}
			else
			{
				cur[j] = dist(x[i], y[j]) + min(prev[j - 1], min(prev[j], cur[j - 1]));
			}
		}
		

	}

	double result = cur[ylen - 1];
	free(cur);
	free(prev);

	return result;
}

